# Paris Housing Analitical Dashboard

## 1. Import dan Persiapan Data

**Import Library**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

### Membaca Dataset

In [2]:
df = pd.read_csv("Paris Housing.csv")

### Eksplorasi Awal Data

In [3]:
df.sample(5)

squareMeters  numberOfRooms  hasYard  hasPool  floors  cityCode  \
2110         84609             59        1        1      29     57704   
871           7812             68        0        1      93     30594   
4325         63657             59        0        0      51     81762   
5849          2536             88        1        0      25     76785   
4339         16839             44        1        0      75     46594   

      cityPartRange  numPrevOwners  made  isNewBuilt  hasStormProtector  \
2110              6              8  2006           1                  0   
871               4              6  2012           0                  1   
4325              7             10  2020           0                  1   
5849             10              5  2018           0                  0   
4339              6              5  1997           0                  1   

      basement  attic  garage  hasStorageRoom  hasGuestRoom      price  \
2110      1459   5381     205               0            10  8469601.6   
871       3559   2197     722               1             0   786725.2   
4325      2036   9364     430               0             0  6367982.0   
5849       264   4785     669               1             0   258795.8   
4339      7765   4838     955               1             0  1692521.7   

     category  
2110   Luxury  
871     Basic  
4325    Basic  
5849    Basic  
4339    Basic

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   squareMeters       10000 non-null  int64  
 1   numberOfRooms      10000 non-null  int64  
 2   hasYard            10000 non-null  int64  
 3   hasPool            10000 non-null  int64  
 4   floors             10000 non-null  int64  
 5   cityCode           10000 non-null  int64  
 6   cityPartRange      10000 non-null  int64  
 7   numPrevOwners      10000 non-null  int64  
 8   made               10000 non-null  int64  
 9   isNewBuilt         10000 non-null  int64  
 10  hasStormProtector  10000 non-null  int64  
 11  basement           10000 non-null  int64  
 12  attic              10000 non-null  int64  
 13  garage             10000 non-null  int64  
 14  hasStorageRoom     10000 non-null  int64  
 15  hasGuestRoom       10000 non-null  int64  
 16  price              10000 non-null 

**Note:**

Berdasarkan df.info():

1. 16 kolom bertipe int64
2. 1 kolom bertipe float64 → price
3. 1 kolom bertipe string → category
4. Tidak terdapat missing values (semua kolom memiliki 10.000 non-null values)

Artinya:

* Dataset dalam kondisi lengkap
* Tidak perlu penanganan missing value

In [5]:
df.describe()

squareMeters  numberOfRooms       hasYard       hasPool        floors  \
count   10000.00000   10000.000000  10000.000000  10000.000000  10000.000000   
mean    49870.13120      50.358400      0.508700      0.496800     50.276300   
std     28774.37535      28.816696      0.499949      0.500015     28.889171   
min        89.00000       1.000000      0.000000      0.000000      1.000000   
25%     25098.50000      25.000000      0.000000      0.000000     25.000000   
50%     50105.50000      50.000000      1.000000      0.000000     50.000000   
75%     74609.75000      75.000000      1.000000      1.000000     76.000000   
max     99999.00000     100.000000      1.000000      1.000000    100.000000   

           cityCode  cityPartRange  numPrevOwners         made    isNewBuilt  \
count  10000.000000   10000.000000   10000.000000  10000.00000  10000.000000   
mean   50225.486100       5.510100       5.521700   2005.48850      0.499100   
std    29006.675799       2.872024       2.856667      9.30809      0.500024   
min        3.000000       1.000000       1.000000   1990.00000      0.000000   
25%    24693.750000       3.000000       3.000000   1997.00000      0.000000   
50%    50693.000000       5.000000       5.000000   2005.50000      0.000000   
75%    75683.250000       8.000000       8.000000   2014.00000      1.000000   
max    99953.000000      10.000000      10.000000   2021.00000      1.000000   

       hasStormProtector      basement        attic       garage  \
count       10000.000000  10000.000000  10000.00000  10000.00000   
mean            0.499900   5033.103900   5028.01060    553.12120   
std             0.500025   2876.729545   2894.33221    262.05017   
min             0.000000      0.000000      1.00000    100.00000   
25%             0.000000   2559.750000   2512.00000    327.75000   
50%             0.000000   5092.500000   5045.00000    554.00000   
75%             1.000000   7511.250000   7540.50000    777.25000   
max             1.000000  10000.000000  10000.00000   1000.00000   

       hasStorageRoom  hasGuestRoom         price  
count    10000.000000   10000.00000  1.000000e+04  
mean         0.503000       4.99460  4.993448e+06  
std          0.500016       3.17641  2.877424e+06  
min          0.000000       0.00000  1.031350e+04  
25%          0.000000       2.00000  2.516402e+06  
50%          1.000000       5.00000  5.016180e+06  
75%          1.000000       8.00000  7.469092e+06  
max          1.000000      10.00000  1.000677e+07

In [6]:
df.describe().to_markdown()

'|       |   squareMeters |   numberOfRooms |      hasYard |      hasPool |     floors |   cityCode |   cityPartRange |   numPrevOwners |        made |   isNewBuilt |   hasStormProtector |   basement |    attic |    garage |   hasStorageRoom |   hasGuestRoom |           price |\n|:------|---------------:|----------------:|-------------:|-------------:|-----------:|-----------:|----------------:|----------------:|------------:|-------------:|--------------------:|-----------:|---------:|----------:|-----------------:|---------------:|----------------:|\n| count |        10000   |      10000      | 10000        | 10000        | 10000      |    10000   |     10000       |     10000       | 10000       | 10000        |        10000        |   10000    | 10000    | 10000     |     10000        |    10000       | 10000           |\n| mean  |        49870.1 |         50.3584 |     0.5087   |     0.4968   |    50.2763 |    50225.5 |         5.5101  |         5.5217  |  2005.49    |     0.4991 

### Pemeriksaan Tipe Data

| Kolom                     | Kondisi Saat Ini | Catatan                           |
| ------------------------- | ---------------- | --------------------------------- |
| `category`                | str              | Sudah benar (categorical)         |
| `hasYard`, `hasPool`, dll | int64            | Seharusnya bisa diubah ke boolean |
| `cityCode`                | int64            | Lebih cocok sebagai kategori      |
| `made`                    | int64            | Tahun bangunan                    |


In [7]:
# Ubah ke boolean
bool_cols = ['hasYard','hasPool','isNewBuilt','hasStormProtector',
             'hasStorageRoom']

df[bool_cols] = df[bool_cols].astype('bool')

# Ubah category
df['category'] = df['category'].astype('category')


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   squareMeters       10000 non-null  int64   
 1   numberOfRooms      10000 non-null  int64   
 2   hasYard            10000 non-null  bool    
 3   hasPool            10000 non-null  bool    
 4   floors             10000 non-null  int64   
 5   cityCode           10000 non-null  int64   
 6   cityPartRange      10000 non-null  int64   
 7   numPrevOwners      10000 non-null  int64   
 8   made               10000 non-null  int64   
 9   isNewBuilt         10000 non-null  bool    
 10  hasStormProtector  10000 non-null  bool    
 11  basement           10000 non-null  int64   
 12  attic              10000 non-null  int64   
 13  garage             10000 non-null  int64   
 14  hasStorageRoom     10000 non-null  bool    
 15  hasGuestRoom       10000 non-null  int64   
 16  price           

### Pemeriksaan Duplikasi

In [9]:
df.duplicated().sum()

np.int64(0)

**Note:**

Tidak terdapat duplikasi data

### Feature Engineering

In [10]:
# Menghitung usia properti berdasarkan tahun pembangunan
df['property_age'] = 2025 - df['made']

# Menjumlahkan total fasilitas tambahan (basement, attic, garage)
df['total_special_rooms'] = df['basement'] + df['attic'] + df['garage']

# Menghitung harga per meter persegi
df['price_per_sqm'] = df['price'] / df['squareMeters']


### Standarisasi Nama Kolom

In [11]:
df = df.rename(columns={
    'squareMeters': 'square_meters',
    'numberOfRooms': 'number_of_rooms',
    'hasYard': 'has_yard',
    'hasPool': 'has_pool',
    'cityCode': 'city_code',
    'cityPartRange': 'city_part_range',
    'numPrevOwners': 'num_prev_owners',
    'made': 'year_built',
    'isNewBuilt': 'is_new_built',
    'hasStormProtector': 'has_storm_protector',
    'hasStorageRoom': 'has_storage_room',
    'hasGuestRoom': 'has_guest_room'
})


### Menambahkan ID

In [12]:
df['id'] = range(1, len(df) + 1)

In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   square_meters        10000 non-null  int64   
 1   number_of_rooms      10000 non-null  int64   
 2   has_yard             10000 non-null  bool    
 3   has_pool             10000 non-null  bool    
 4   floors               10000 non-null  int64   
 5   city_code            10000 non-null  int64   
 6   city_part_range      10000 non-null  int64   
 7   num_prev_owners      10000 non-null  int64   
 8   year_built           10000 non-null  int64   
 9   is_new_built         10000 non-null  bool    
 10  has_storm_protector  10000 non-null  bool    
 11  basement             10000 non-null  int64   
 12  attic                10000 non-null  int64   
 13  garage               10000 non-null  int64   
 14  has_storage_room     10000 non-null  bool    
 15  has_guest_room       10000 non-

### Outlier Checking

In [14]:
num_cols = df.select_dtypes(include=['int64','float64']).columns
num_cols

Index(['square_meters', 'number_of_rooms', 'floors', 'city_code',
       'city_part_range', 'num_prev_owners', 'year_built', 'basement', 'attic',
       'garage', 'has_guest_room', 'price', 'property_age',
       'total_special_rooms', 'price_per_sqm', 'id'],
      dtype='str')

In [15]:
# Check Outlier using IQR
def detect_outliers_iqr(data, column):
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    outliers = data[(data[column] < lower) | (data[column] > upper)]
    
    return outliers.shape[0]

for col in num_cols:
    print(col, ":", detect_outliers_iqr(df, col))


square_meters : 0
number_of_rooms : 0
floors : 0
city_code : 0
city_part_range : 0
num_prev_owners : 0
year_built : 0
basement : 0
attic : 0
garage : 0
has_guest_room : 0
price : 0
property_age : 0
total_special_rooms : 0
price_per_sqm : 1264
id : 0


**Note:**

  Terdapat outlier pada kolom baru `price_per_sqm` dengan jumlah 1264

In [16]:
"""
Capping (atau disebut juga winsorizing) adalah teknik menangani outlier dengan cara:

1. Mengganti nilai ekstrem dengan batas maksimum atau minimum tertentu
2. BUKAN menghapus datanya

Jadi datanya tetap ada, hanya nilainya yang “dipotong” ke batas wajar.
"""

def cap_outliers(data, column):
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    data[column] = data[column].clip(lower, upper)
    
for col in num_cols:
    cap_outliers(df, col)

In [17]:
for col in num_cols:
    print(col, ":", detect_outliers_iqr(df, col))

square_meters : 0
number_of_rooms : 0
floors : 0
city_code : 0
city_part_range : 0
num_prev_owners : 0
year_built : 0
basement : 0
attic : 0
garage : 0
has_guest_room : 0
price : 0
property_age : 0
total_special_rooms : 0
price_per_sqm : 0
id : 0


In [18]:
df

square_meters  number_of_rooms  has_yard  has_pool  floors  city_code  \
0             75523                3     False      True      63       9373   
1             80771               39      True      True      98      39381   
2             55712               58     False      True      19      34457   
3             32316               47     False     False       6      27939   
4             70429               19      True      True      90      38045   
...             ...              ...       ...       ...     ...        ...   
9995           1726               89     False      True       5      73133   
9996          44403               29      True      True      12      34606   
9997          83841                3     False     False      69      80933   
9998          59036               70     False     False      96      55856   
9999           1440               84     False     False      49      18412   

      city_part_range  num_prev_owners  year_built  is_new_built  ...  attic  \
0                   3                8        2005         False  ...   9005   
1                   8                6        2015          True  ...   2436   
2                   6                8        2021         False  ...   8852   
3                  10                4        2012         False  ...   7141   
4                   3                7        1990          True  ...   2429   
...               ...              ...         ...           ...  ...    ...   
9995                7                6        2009         False  ...   1698   
9996                9                4        1990         False  ...   1742   
9997               10               10        2005          True  ...   7730   
9998                1                3        2010         False  ...   6174   
9999                6               10        1994          True  ...   2024   

      garage  has_storage_room  has_guest_room      price  category  \
0        956             False               7  7559081.5     Basic   
1        128              True               2  8085989.5    Luxury   
2        135              True               9  5574642.1     Basic   
3        359             False               3  3232561.2     Basic   
4        292              True               4  7055052.0    Luxury   
...      ...               ...             ...        ...       ...   
9995     218             False               4   176425.9     Basic   
9996     230             False               0  4448474.0     Basic   
9997     345              True               9  8390030.5     Basic   
9998     339              True               4  5905107.0     Basic   
9999     278              True               6   146708.4     Basic   

      property_age total_special_rooms  price_per_sqm     id  
0               20               14274     100.089794      1  
1               10                6217     100.110058      2  
2                4               11924     100.061784      3  
3               13                8159     100.029744      4  
4               35               11156     100.172543      5  
...            ...                 ...            ...    ...  
9995            16               11227     100.532565   9996  
9996            35               11033     100.184087   9997  
9997            20               16379     100.070735   9998  
9998            15                9103     100.025527   9999  
9999            31               10787     100.532565  10000  

[10000 rows x 22 columns]

### 7. Export The Result

In [19]:
df.to_csv("paris_housing_cleaned.csv", index=False)

In [20]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   square_meters        10000 non-null  int64   
 1   number_of_rooms      10000 non-null  int64   
 2   has_yard             10000 non-null  bool    
 3   has_pool             10000 non-null  bool    
 4   floors               10000 non-null  int64   
 5   city_code            10000 non-null  int64   
 6   city_part_range      10000 non-null  int64   
 7   num_prev_owners      10000 non-null  int64   
 8   year_built           10000 non-null  int64   
 9   is_new_built         10000 non-null  bool    
 10  has_storm_protector  10000 non-null  bool    
 11  basement             10000 non-null  int64   
 12  attic                10000 non-null  int64   
 13  garage               10000 non-null  int64   
 14  has_storage_room     10000 non-null  bool    
 15  has_guest_room       10000 non-

In [22]:
categorical_cols = df.select_dtypes(include=['object', 'int64', 'bool']).columns

for col in categorical_cols:
    print(f"\nValue counts untuk kolom: {col}")
    print(df[col].value_counts())



Value counts untuk kolom: square_meters
square_meters
79770    3
33749    3
96526    3
16006    3
68985    3
        ..
21514    1
1726     1
44403    1
83841    1
59036    1
Name: count, Length: 9483, dtype: int64

Value counts untuk kolom: number_of_rooms
number_of_rooms
54    129
4     120
22    119
47    118
3     116
     ... 
6      85
31     84
34     84
40     82
9      75
Name: count, Length: 100, dtype: int64

Value counts untuk kolom: has_yard
has_yard
True     5087
False    4913
Name: count, dtype: int64

Value counts untuk kolom: has_pool
has_pool
False    5032
True     4968
Name: count, dtype: int64

Value counts untuk kolom: floors
floors
97     126
55     122
77     117
3      116
28     116
      ... 
15      83
74      83
48      83
100     82
92      75
Name: count, Length: 100, dtype: int64

Value counts untuk kolom: city_code
city_code
82521    3
79444    3
83194    3
95054    3
36929    3
        ..
73133    1
34606    1
80933    1
55856    1
18412    1
Name: cou